In [ ]:
# CT image is input

# INPUT: CT Images is input
# OUTPUT: RP file
# PROCESS: algorith predict by each slice. make_lines_process. wrap to RP file by RP file template

# Template processing note: 


# Some Filepath for RP 


In [ ]:
#rp_info =dicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')
#rp_info = pydicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')

# Template Data
rp_info = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')


# Only Read File

In [36]:

# Start research from here

import pydicom

#rp_info =dicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')
#rp_info = pydicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')

# Template Data
rp_info = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')

App = rp_info.ApplicationSetupSequence
ChannelSequence = App[0].ChannelSequence
#print('ChannelSequence[0].BrachyControlPointSequence = ')
#print(ChannelSequence[0].BrachyControlPointSequence)
print('ChannelSequence[0] = ')

for idx in range(len(ChannelSequence)):
    cs = ChannelSequence[idx]
    print('idx = {} ->'.format(idx))
    print(cs)
    break
    

for idx, cp in enumerate(ChannelSequence[0].BrachyControlPointSequence):
    
    #print(cp)
    #print("position:", cp.ControlPoint3DPosition, end = " ")
    #print("position:", cp.ControlPoint3DPosition)
    #print(type(cp.ControlPoint3DPosition))
    
    if (idx >= 0 ) :
        print('idx == {} -> cp = '.format(idx))
        print(cp)
        print(type(cp.ControlPointIndex))
        print('===============')
    
    #continue
    print("Time:", cp.CumulativeTimeWeight)
    try: 
        for jj in cp.BrachyReferencedDoseReferenceSequence:
            #print(jj)
            print("ref:", rp_info.DoseReferenceSequence[jj.ReferencedDoseReferenceNumber-1].DoseReferenceDescription , end = " ")
            print("coeff:", jj.CumulativeDoseReferenceCoefficient)
    except Exception as e:
        print(e)
        pass
    print('')

ChannelSequence[0] = 
idx = 0 ->
(3006, 0084) Referenced ROI Number               IS: "18"
(300a, 0110) Number of Control Points            IS: "38"
(300a, 0282) Channel Number                      IS: "1"
(300a, 0284) Channel Length                      DS: "1300"
(300a, 0286) Channel Total Time                  DS: "0"
(300a, 0288) Source Movement Type                CS: 'STEPWISE'
(300a, 0290) Source Applicator Number            IS: "1"
(300a, 0291) Source Applicator ID                SH: 'Tandom_2'
(300a, 0292) Source Applicator Type              CS: 'RIGID'
(300a, 0294) Source Applicator Name              LO: ''
(300a, 0296) Source Applicator Length            DS: "1300"
(300a, 02a0) Source Applicator Step Size         DS: "5"
(300a, 02a2) Transfer Tube Number                IS: ''
(300a, 02c8) Final Cumulative Time Weight        DS: "0"
(300a, 02d0)  Brachy Control Point Sequence   38 item(s) ---- 
   (300a, 0112) Control Point Index                 IS: "0"
   (300a, 02d2) Contro

In [40]:
ChannelSequence[3]

IndexError: list index out of range

# Write File and read file again 

In [25]:

# Write File Test and read again 

import pydicom

#rp_info =dicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')
#rp_info = pydicom.read_file(r'D:\kai\DICOMtoNPZ\ral\RP.1.2.246.352.71.5.417454940236.2035996.20190820095628.dcm')

# Template Data
rp_info = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')
out_rp_filepath = r'Brachy.RP.out.dcm'

App = rp_info.ApplicationSetupSequence
ChannelSequence = App[0].ChannelSequence
print('ChannelSequence[0] = ')

for idx, cp in enumerate(ChannelSequence[0].BrachyControlPointSequence):
    print("position:", cp.ControlPoint3DPosition)
    if (idx >= 0 ) :
        #print('idx == {} -> cp = '.format(idx))
        #print(cp)
        #print('===============')
        pass
    if (idx == 0) :
        cp.ControlPoint3DPosition[0] = 33.33333
    
        
    
    #continue
    print("Time:", cp.CumulativeTimeWeight)
    try: 
        for jj in cp.BrachyReferencedDoseReferenceSequence:
            #print(jj)
            print("ref:", rp_info.DoseReferenceSequence[jj.ReferencedDoseReferenceNumber-1].DoseReferenceDescription , end = " ")
            print("coeff:", jj.CumulativeDoseReferenceCoefficient)
    except Exception as e:
        print(e)
        pass
    print('')
pydicom.write_file(out_rp_filepath, rp_info)


in_rp_filepath = r'Brachy.RP.out.dcm'
rp_fp = pydicom.read_file(in_rp_filepath)
seq = rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence
for idx, cp in enumerate(seq):
    print("position:", cp.ControlPoint3DPosition)


ChannelSequence[0] = 
position: ['12.778672628366', '-14.486222523329', '1.80953906063796']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.778672628366', '-14.486222523329', '1.80953906063796']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.4938084644467', '-12.702280882149', '-8.025927609987']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.4938084644467', '-12.702280882149', '-8.025927609987']
Time: 0
'Dataset' object has no attribute 'BrachyReferencedDoseReferenceSequence'

position: ['12.351376382487', '-11.810310061558', '-12.943

# The format is like 
rp_info.ApplicationSetupSequence[0].ChannelSequence[0].NumberOfControlPoints is 38
rp_fp.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence[ 0 .. 37]



(300a, 0112) Control Point Index                 IS: "0"
(300a, 02d2) Control Point Relative Position     DS: "3.5"
(300a, 02d4) Control Point 3D Position           DS: ['12.778672628366', '-14.486222523329', '1.80953906063796']
(300a, 02d6) Cumulative Time Weight              DS: "0"

(300a, 0112) Control Point Index                 IS: "1"
(300a, 02d2) Control Point Relative Position     DS: "3.5"
(300a, 02d4) Control Point 3D Position           DS: ['12.778672628366', '-14.486222523329', '1.80953906063796']
(300a, 02d6) Cumulative Time Weight              DS: "0"


(300a, 0112) Control Point Index                 IS: "2"
(300a, 02d2) Control Point Relative Position     DS: "8.5"
(300a, 02d4) Control Point 3D Position           DS: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
(300a, 02d6) Cumulative Time Weight              DS: "0"


(300a, 0112) Control Point Index                 IS: "3"
(300a, 02d2) Control Point Relative Position     DS: "8.5"
(300a, 02d4) Control Point 3D Position           DS: ['12.6362405464063', '-13.594251702739', '-3.1081942746745']
(300a, 02d6) Cumulative Time Weight              DS: "0"

...

(300a, 0112) Control Point Index                 IS: "36"
(300a, 02d2) Control Point Relative Position     DS: "93.5"
(300a, 02d4) Control Point 3D Position           DS: ['10.2148951530921', '1.56925224729876', '-86.709660974987']
(300a, 02d6) Cumulative Time Weight              DS: "0"


(300a, 0112) Control Point Index                 IS: "37"
(300a, 02d2) Control Point Relative Position     DS: "93.5"
(300a, 02d4) Control Point 3D Position           DS: ['10.2148951530921', '1.56925224729876', '-86.709660974987']
(300a, 02d6) Cumulative Time Weight              DS: "0"

In [76]:
import pydicom
import copy

rp_info = pydicom.read_file(r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm')

#App = rp_info.ApplicationSetupSequence
#ChannelSequence = App[0].ChannelSequence.Number of Control Points

print(rp_info.ApplicationSetupSequence[0].ChannelSequence[0].NumberOfControlPoints)
#print(rp_info.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence)

# We call BrachyControlPointSequence[i] BCPItem here
BCPItemTemplate = copy.deepcopy(rp_info.ApplicationSetupSequence[0].ChannelSequence[0].BrachyControlPointSequence[37])
#print(BrachyControlPointSequenceItemTemplate)

#
# (ControlPointIndex, ControlPointRelativePosition, ControlPoint3DPosition, CumulativeTimeWeight)
# (0, 3.5, [,,], 0 )
# (0, )

#for idx, cp in enumerate(ChannelSequence[0].BrachyControlPointSequence):

print(rp_info)

38
(0008, 0005) Specific Character Set              CS: 'ISO_IR 192'
(0008, 0012) Instance Creation Date              DA: '20191015'
(0008, 0013) Instance Creation Time              TM: '164400.648000'
(0008, 0016) SOP Class UID                       UI: RT Plan Storage
(0008, 0018) SOP Instance UID                    UI: 1.2.246.352.71.5.417454940236.2063186.20191015164204
(0008, 0020) Study Date                          DA: '20191015'
(0008, 0030) Study Time                          TM: '103307.760000'
(0008, 0050) Accession Number                    SH: '2019-10-15-10:36'
(0008, 0060) Modality                            CS: 'RTPLAN'
(0008, 0070) Manufacturer                        LO: 'Varian Medical Systems'
(0008, 0090) Referring Physician's Name          PN: ''
(0008, 1010) Station Name                        SH: 'ARIA136SQL'
(0008, 1030) Study Description                   LO: 'Pelvis^3_IC (Adult)'
(0008, 103e) Series Description                  LO: 'ARIA RadOnc Plans'
(0008, 1

## Write related RS Data in RP Tamplate

### OperatorsName 
cylin is in template. But Who is the people ?

### PhysiciansOfRecord
Just copy from RS

### FrameOfReferenceUID
In RS, ReferencedFrameOfReferenceSequence[0].FrameOfReferenceUID


### Basic thing copy from RS
PhysiciansOfRecord, PatientName, PatientID, PatientBirthDate, PatientBirthTime, PatientSex,
DeviceSerialNumber, SoftwareVersions, StudyID, StudyDate, StudyTime, StudyInstanceUID

### ReferencedStructureSetSequence
In RP,   
ReferencedStructureSetSequence[0].ReferencedSOPClassUID = 1.2.840.10008.5.1.4.1.1.481.3  ReferencedStructureSetSequence[0].ReferencedSOPInstanceUID =   1.2.246.352.71.4.417454940236.287582.20191015164114  
So,   
RP ReferencedStructureSetSequence[0].ReferencedSOPClassUID =  RS SOPClassUID  
RP ReferencedStructureSetSequence[0].ReferencedSOPInstanceUID =  RS SOPInstanceUID  

### InstanceCreation
RS InstanceCreationDate = 20191015  
RS InstanceCreationTime = 164359.971000  
RP InstanceCreationDate = 20191015  
RP InstanceCreationTime = 164400.648000  
RP RTPlanDate = 20191015  
RP RTPlanTime = 164203.870000   (What if compare to RP InstanceCreationTime ?)    
RS StudyDate = 20191015   
RS StudyTime = 103307.760000  
  
Conclusion,     
StudyDate, InstanceCreationDate are all the same in RS and RP.    
RS StudyTime < RP RTPlanTime < RS InstanceCreationTime < RP InstanceCreationTime   



# Implemnet write procedure

In [73]:
import pydicom
import os
rp_template_filepath = r'RP_Template/Brachy_RP.1.2.246.352.71.5.417454940236.2063186.20191015164204.dcm'
folder = r'RAL_plan_new_20190905/29059811-1'

# Set rs_filepath and ct_filelist, which are INPUT
rs_filepath = ''
ct_filelist = []
for file in os.listdir(folder):
    filepath = os.path.join(folder, file)
    fp = pydicom.read_file(filepath)
    if (fp.Modality == 'CT'):
        ct_filelist.append(filepath)
    elif (fp.Modality == 'RTSTRUCT'):
        rs_filepath = filepath
        
# Read RS file as input
rs_fp = pydicom.read_file(rs_filepath)
# read RP tempalte into rp_fp
rp_fp = pydicom.read_file(rp_template_filepath)
out_rp_filepath = r'out.brachy.rp.dcm'


rp_fp.OperatorsName = 'cylin'
rp_fp.PhysiciansOfRecord = rs_fp.PhysiciansOfRecord
rp_fp.FrameOfReferenceUID = rs_fp.ReferencedFrameOfReferenceSequence[0].FrameOfReferenceUID
rp_fp.ReferencedStructureSetSequence[0].ReferencedSOPClassUID =  rs_fp.SOPClassUID  
rp_fp.ReferencedStructureSetSequence[0].ReferencedSOPInstanceUID = rs_fp.SOPInstanceUID  

directAttrSet = [
    'PhysiciansOfRecord','PatientName', 'PatientID', 
    'PatientBirthDate', 'PatientBirthTime', 'PatientSex', 
    'DeviceSerialNumber', 'SoftwareVersions', 'StudyID', 
    'StudyDate', 'StudyTime', 'StudyInstanceUID']
for attr in directAttrSet:
    val = getattr(rs_fp, attr)
    setattr(rp_fp, attr, val)


# 
# to satisfy the condition 
# (1) RS InstanceCreationDate == RP InstanceCreationDate == RP RTPlanDate == RS StudyDate
# (2) RS StudyTime < RP RTPlanTime < RS InstanceCreationTime < RP InstanceCreationTime   
# for (1)
rp_fp.InstanceCreationDate = rp_fp.RTPlanDate = rp_fp.StudyDate = rs_fp.StudyDate
# for (2)
#print('type(rs_fp.StudyTime) = {}'.format(type(rs_fp.StudyTime))) # type is str
#print('type(rp_fp.RTPlanTime) = {}'.format(type(rp_fp.RTPlanTime))) # type is str
rp_fp.RTPlanTime= str(float(rs_fp.StudyTime) + 0.001)
rp_fp.InstanceCreationTime = str(float(rs_fp.InstanceCreationTime) + 0.001)


pydicom.write_file(out_rp_filepath, rp_fp)



# Check the data in output RP filepath 

In [74]:
rp_fp = pydicom.read_file(out_rp_filepath)
print(rp_fp)

(0008, 0005) Specific Character Set              CS: 'ISO_IR 192'
(0008, 0012) Instance Creation Date              DA: '20190903'
(0008, 0013) Instance Creation Time              TM: '134930.715'
(0008, 0016) SOP Class UID                       UI: RT Plan Storage
(0008, 0018) SOP Instance UID                    UI: 1.2.246.352.71.5.417454940236.2063186.20191015164204
(0008, 0020) Study Date                          DA: '20190903'
(0008, 0030) Study Time                          TM: '091942.790000'
(0008, 0050) Accession Number                    SH: '2019-10-15-10:36'
(0008, 0060) Modality                            CS: 'RTPLAN'
(0008, 0070) Manufacturer                        LO: 'Varian Medical Systems'
(0008, 0090) Referring Physician's Name          PN: ''
(0008, 1010) Station Name                        SH: 'ARIA136SQL'
(0008, 1030) Study Description                   LO: 'Pelvis^3_IC (Adult)'
(0008, 103e) Series Description                  LO: 'ARIA RadOnc Plans'
(0008, 1048) P

In [75]:
print(rs_fp)

(0008, 0005) Specific Character Set              CS: 'ISO_IR 192'
(0008, 0012) Instance Creation Date              DA: '20190905'
(0008, 0013) Instance Creation Time              TM: '134930.714000'
(0008, 0016) SOP Class UID                       UI: RT Structure Set Storage
(0008, 0018) SOP Instance UID                    UI: 1.2.246.352.71.4.417454940236.283407.20190903102804
(0008, 0020) Study Date                          DA: '20190903'
(0008, 0030) Study Time                          TM: '091942.790000'
(0008, 0050) Accession Number                    SH: '2019-09-03-09:33'
(0008, 0060) Modality                            CS: 'RTSTRUCT'
(0008, 0070) Manufacturer                        LO: 'Varian Medical Systems'
(0008, 0090) Referring Physician's Name          PN: ''
(0008, 1010) Station Name                        SH: 'ARIA136SQL'
(0008, 1030) Study Description                   LO: 'Pelvis^3_IC (Adult)'
(0008, 103e) Series Description                  LO: 'ARIA RadOnc Structur